File: src/infrastructure/persistence/filesystem_doc_scanner.py
"""Filesystem adapter for document health scanning.

FilesystemDocScanner implements DocScannerProtocol by reading TOML registries,
checking file existence, and parsing YAML frontmatter for last_reviewed dates.
Uses simple regex parsing -- no pyyaml dependency required.
"""

from __future__ import annotations

import re
from datetime import date
from typing import TYPE_CHECKING

from src.domain.models.doc_health import (
    DocRegistryEntry,
    DocStatus,
    create_doc_status,
)

if TYPE_CHECKING:
    from pathlib import Path

_FRONTMATTER_RE = re.compile(r"^---\s*\n(.*?)\n---", re.DOTALL)
_LAST_REVIEWED_RE = re.compile(r"^last_reviewed:\s*(.+)$", re.MULTILINE)
_PLACEHOLDER_RE = re.compile(r"^[A-Z]{4}-[A-Z]{2}-[A-Z]{2}$")


def _parse_last_reviewed(content: str) -> date | None:
    """Extract last_reviewed date from markdown file content.

    Looks for YAML frontmatter delimited by --- and extracts the
    last_reviewed field. Returns None if not found, if the value is
    a placeholder like YYYY-MM-DD, or if the date is invalid.

    Args:
        content: Full text content of a markdown file.

    Returns:
        Parsed date, or None if unavailable.
    """
    fm_match = _FRONTMATTER_RE.search(content)
    if not fm_match:
        return None

    frontmatter = fm_match.group(1)
    lr_match = _LAST_REVIEWED_RE.search(frontmatter)
    if not lr_match:
        return None

    raw = lr_match.group(1).strip().strip('"').strip("'")

    if _PLACEHOLDER_RE.match(raw):
        return None

    try:
        return date.fromisoformat(raw)
    except ValueError:
        return None


def _parse_toml_value(value: str) -> str | int:
    """Parse a single TOML value (string or integer).

    Args:
        value: Stripped TOML value string.

    Returns:
        Unquoted string or parsed integer.
    """
    if (value.startswith('"') and value.endswith('"')) or (
        value.startswith("'") and value.endswith("'")
    ):
        return value[1:-1]
    try:
        return int(value)
    except ValueError:
        return value


def _parse_toml_docs(content: str) -> list[dict[str, str | int]]:
    """Parse [[docs]] entries from a simple TOML file.

    This is a minimal parser that handles the specific format used by
    doc-registry.toml. It does not handle all TOML features.

    Args:
        content: TOML file content.

    Returns:
        List of dicts with string/int values per entry.
    """
    entries: list[dict[str, str | int]] = []
    current: dict[str, str | int] | None = None

    for line in content.splitlines():
        stripped = line.strip()
        if stripped == "[[docs]]":
            if current is not None:
                entries.append(current)
            current = {}
            continue

        if current is None or not stripped or stripped.startswith("#") or "=" not in stripped:
            continue

        key, _, value = stripped.partition("=")
        current[key.strip()] = _parse_toml_value(value.strip())

    if current is not None:
        entries.append(current)

    return entries


class FilesystemDocScanner:
    """Scans the filesystem for document health status.

    Implements DocScannerProtocol by reading TOML registries,
    checking file existence, and parsing YAML frontmatter.
    """

    def load_registry(self, registry_path: Path) -> tuple[DocRegistryEntry, ...]:
        """Load document registry entries from a TOML file.

        Args:
            registry_path: Path to the doc-registry.toml file.

        Returns:
            Tuple of DocRegistryEntry, empty if file is missing or broken.
        """
        if not registry_path.is_file():
            return ()

        try:
            content = registry_path.read_text()
        except OSError:
            return ()

        raw_entries = _parse_toml_docs(content)
        entries: list[DocRegistryEntry] = []
        for raw in raw_entries:
            path = raw.get("path")
            if not isinstance(path, str):
                continue
            owner = raw.get("owner")
            owner_str = str(owner) if owner is not None else None
            interval = raw.get("review_interval_days", 30)
            interval_int = int(interval) if not isinstance(interval, int) else interval
            entries.append(
                DocRegistryEntry(
                    path=path,
                    owner=owner_str,
                    review_interval_days=interval_int,
                )
            )

        return tuple(entries)

    def scan_registered(
        self,
        entries: tuple[DocRegistryEntry, ...],
        project_dir: Path,
    ) -> tuple[DocStatus, ...]:
        """Scan registered documents for health status.

        For each entry, checks file existence and parses frontmatter
        for the last_reviewed date. Uses create_doc_status factory.

        Args:
            entries: Registry entries to scan.
            project_dir: Project root directory.

        Returns:
            Tuple of DocStatus for each registered document.
        """
        statuses: list[DocStatus] = []
        for entry in entries:
            file_path = project_dir / entry.path
            exists = file_path.is_file()

            last_reviewed: date | None = None
            if exists:
                try:
                    content = file_path.read_text()
                    last_reviewed = _parse_last_reviewed(content)
                except OSError:
                    pass

            statuses.append(
                create_doc_status(
                    path=entry.path,
                    exists=exists,
                    last_reviewed=last_reviewed,
                    review_interval_days=entry.review_interval_days,
                    owner=entry.owner,
                )
            )

        return tuple(statuses)

    def scan_unregistered(
        self,
        docs_dir: Path,
        registered_paths: frozenset[str],
        exclude_dirs: tuple[str, ...] = ("templates", "beads_templates"),
    ) -> tuple[DocStatus, ...]:
        """Scan for markdown files not in the registry.

        Globs for *.md files, skips registered paths and excluded dirs,
        and checks frontmatter presence.

        Args:
            docs_dir: Directory to scan for .md files.
            registered_paths: Paths already tracked in the registry.
            exclude_dirs: Directory names to skip.

        Returns:
            Tuple of DocStatus for unregistered documents.
        """
        if not docs_dir.is_dir():
            return ()

        statuses: list[DocStatus] = []
        for md_file in sorted(docs_dir.rglob("*.md")):
            # Build relative path from docs_dir's parent (project root)
            try:
                rel_path = str(md_file.relative_to(docs_dir.parent))
            except ValueError:
                rel_path = str(md_file.relative_to(docs_dir))

            # Skip registered paths
            if rel_path in registered_paths:
                continue

            # Skip excluded directories
            parts = md_file.relative_to(docs_dir).parts
            if any(part in exclude_dirs for part in parts):
                continue

            # Check frontmatter
            try:
                content = md_file.read_text()
                last_reviewed = _parse_last_reviewed(content)
            except OSError:
                last_reviewed = None

            statuses.append(
                create_doc_status(
                    path=rel_path,
                    exists=True,
                    last_reviewed=last_reviewed,
                )
            )

        return tuple(statuses)

File: internal/dochealth/infrastructure/filesystem_doc_scanner.go
// Package infrastructure provides adapters for the DocHealth bounded context.
package infrastructure

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	dochealthdomain "github.com/alty-cli/alty/internal/dochealth/domain"
)

var (
	frontmatterRE  = regexp.MustCompile(`(?s)^---\s*\n(.*?)\n---`)
	lastReviewedRE = regexp.MustCompile(`(?m)^last_reviewed:\s*(.+)$`)
	placeholderRE  = regexp.MustCompile(`^[A-Z]{4}-[A-Z]{2}-[A-Z]{2}$`)
	markdownLinkRE = regexp.MustCompile(`(?:^|[^!])\[([^\]]*)\]\(([^)]*)\)`)
)

// FilesystemDocScanner scans the filesystem for document health status.
type FilesystemDocScanner struct{}

// NewFilesystemDocScanner creates a new FilesystemDocScanner.
func NewFilesystemDocScanner() *FilesystemDocScanner {
	return &FilesystemDocScanner{}
}

// LoadRegistry loads document registry entries from a TOML file.
func (s *FilesystemDocScanner) LoadRegistry(registryPath string) ([]dochealthdomain.DocRegistryEntry, error) {
	if _, err := os.Stat(registryPath); os.IsNotExist(err) {
		return nil, nil
	}

	content, err := os.ReadFile(registryPath)
	if err != nil {
		return nil, nil
	}

	rawEntries := parseTOMLDocs(string(content))
	var entries []dochealthdomain.DocRegistryEntry
	for _, raw := range rawEntries {
		path, ok := raw["path"].(string)
		if !ok {
			continue
		}
		owner := ""
		if o, ok := raw["owner"].(string); ok {
			owner = o
		}
		interval := 30
		if i, ok := raw["review_interval_days"].(int); ok {
			interval = i
		}
		entry, err := dochealthdomain.NewDocRegistryEntry(path, owner, interval)
		if err != nil {
			continue
		}
		entries = append(entries, entry)
	}
	return entries, nil
}

// ScanRegistered scans registered documents for health status.
func (s *FilesystemDocScanner) ScanRegistered(entries []dochealthdomain.DocRegistryEntry, projectDir string) ([]dochealthdomain.DocStatus, error) {
	var statuses []dochealthdomain.DocStatus
	now := time.Now().Truncate(24 * time.Hour)

	for _, entry := range entries {
		filePath := filepath.Join(projectDir, entry.Path())
		info, err := os.Stat(filePath)
		exists := err == nil && !info.IsDir()

		var lastReviewed *time.Time
		var brokenLinks []dochealthdomain.BrokenLink
		if exists {
			content, readErr := os.ReadFile(filePath)
			if readErr == nil {
				lastReviewed = parseLastReviewed(string(content))
				links := extractMarkdownLinks(string(content))
				brokenLinks = checkBrokenLinks(links, filePath)
			}
		}

		status := dochealthdomain.CreateDocStatus(
			entry.Path(),
			exists,
			lastReviewed,
			entry.ReviewIntervalDays(),
			entry.Owner(),
			&now,
			brokenLinks,
		)
		statuses = append(statuses, status)
	}
	return statuses, nil
}

// ScanUnregistered scans for markdown files not in the registry.
func (s *FilesystemDocScanner) ScanUnregistered(
	docsDir string,
	registeredPaths map[string]bool,
	excludeDirs []string,
) ([]dochealthdomain.DocStatus, error) {
	if _, err := os.Stat(docsDir); os.IsNotExist(err) {
		return nil, nil
	}

	now := time.Now().Truncate(24 * time.Hour)
	excludeSet := make(map[string]bool)
	for _, d := range excludeDirs {
		excludeSet[d] = true
	}

	var mdFiles []string
	err := filepath.Walk(docsDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() {
			return nil
		}
		if filepath.Ext(path) == ".md" {
			mdFiles = append(mdFiles, path)
		}
		return nil
	})
	if err != nil {
		return nil, fmt.Errorf("walking docs directory: %w", err)
	}
	sort.Strings(mdFiles)

	var statuses []dochealthdomain.DocStatus
	parentDir := filepath.Dir(docsDir)

	for _, mdFile := range mdFiles {
		// Build relative path from docs_dir's parent (project root)
		relPath, err := filepath.Rel(parentDir, mdFile)
		if err != nil {
			relPath, _ = filepath.Rel(docsDir, mdFile)
		}

		// Skip registered paths
		if registeredPaths != nil && registeredPaths[relPath] {
			continue
		}

		// Skip excluded directories
		relFromDocs, _ := filepath.Rel(docsDir, mdFile)
		parts := strings.Split(relFromDocs, string(filepath.Separator))
		excluded := false
		for _, part := range parts {
			if excludeSet[part] {
				excluded = true
				break
			}
		}
		if excluded {
			continue
		}

		// Check frontmatter and links
		content, readErr := os.ReadFile(mdFile)
		var lastReviewed *time.Time
		var brokenLinks []dochealthdomain.BrokenLink
		if readErr == nil {
			lastReviewed = parseLastReviewed(string(content))
			links := extractMarkdownLinks(string(content))
			brokenLinks = checkBrokenLinks(links, mdFile)
		}

		status := dochealthdomain.CreateDocStatus(
			relPath,
			true,
			lastReviewed,
			30,
			"",
			&now,
			brokenLinks,
		)
		statuses = append(statuses, status)
	}
	return statuses, nil
}

func parseLastReviewed(content string) *time.Time {
	fmMatch := frontmatterRE.FindStringSubmatch(content)
	if fmMatch == nil {
		return nil
	}

	lrMatch := lastReviewedRE.FindStringSubmatch(fmMatch[1])
	if lrMatch == nil {
		return nil
	}

	raw := strings.TrimSpace(lrMatch[1])
	raw = strings.Trim(raw, "\"'")

	if placeholderRE.MatchString(raw) {
		return nil
	}

	t, err := time.Parse("2006-01-02", raw)
	if err != nil {
		return nil
	}
	return &t
}

type linkInfo struct {
	lineNumber int
	linkText   string
	target     string
}

func extractMarkdownLinks(content string) []linkInfo {
	var results []linkInfo
	inFence := false
	for lineno, line := range strings.Split(content, "\n") {
		stripped := strings.TrimSpace(line)
		if strings.HasPrefix(stripped, "```") || strings.HasPrefix(stripped, "~~~") {
			inFence = !inFence
			continue
		}
		if inFence {
			continue
		}
		matches := markdownLinkRE.FindAllStringSubmatch(line, -1)
		for _, m := range matches {
			results = append(results, linkInfo{
				lineNumber: lineno + 1,
				linkText:   m[1],
				target:     m[2],
			})
		}
	}
	return results
}

func checkBrokenLinks(links []linkInfo, docPath string) []dochealthdomain.BrokenLink {
	var broken []dochealthdomain.BrokenLink
	for _, link := range links {
		if link.target == "" {
			bl, err := dochealthdomain.NewBrokenLink(link.lineNumber, link.linkText, link.target, "empty target")
			if err == nil {
				broken = append(broken, bl)
			}
			continue
		}
		if strings.HasPrefix(link.target, "http://") || strings.HasPrefix(link.target, "https://") ||
			strings.HasPrefix(link.target, "mailto:") {
			continue
		}
		if strings.HasPrefix(link.target, "#") {
			continue
		}
		pathPart := strings.SplitN(link.target, "#", 2)[0]
		if pathPart == "" {
			continue
		}
		resolved := filepath.Join(filepath.Dir(docPath), pathPart)
		if _, err := os.Stat(resolved); os.IsNotExist(err) {
			bl, err := dochealthdomain.NewBrokenLink(link.lineNumber, link.linkText, link.target, "target not found")
			if err == nil {
				broken = append(broken, bl)
			}
		}
	}
	return broken
}

// parseTOMLDocs parses [[docs]] entries from a simple TOML file.
func parseTOMLDocs(content string) []map[string]any {
	var entries []map[string]any
	var current map[string]any

	for _, line := range strings.Split(content, "\n") {
		stripped := strings.TrimSpace(line)
		if stripped == "[[docs]]" {
			if current != nil {
				entries = append(entries, current)
			}
			current = make(map[string]any)
			continue
		}
		if current == nil || stripped == "" || strings.HasPrefix(stripped, "#") || !strings.Contains(stripped, "=") {
			continue
		}
		key, value, found := strings.Cut(stripped, "=")
		if !found {
			continue
		}
		k := strings.TrimSpace(key)
		v := strings.TrimSpace(value)
		current[k] = parseTOMLValue(v)
	}
	if current != nil {
		entries = append(entries, current)
	}
	return entries
}

func parseTOMLValue(value string) any {
	if (strings.HasPrefix(value, "\"") && strings.HasSuffix(value, "\"")) ||
		(strings.HasPrefix(value, "'") && strings.HasSuffix(value, "'")) {
		return value[1 : len(value)-1]
	}
	var n int
	if _, err := fmt.Sscanf(value, "%d", &n); err == nil {
		return n
	}
	return value
}
